ocaml/xapi/extauth_plugin_ADpbis.ml

```diff
@@ -734,83 +734,94 @@ module AuthADlw : Auth_signature.AUTH_MODULE = struct
-               Not_found, waiting 5 secs to try again"
-              i max ;
-            Thread.delay 5.0 ;
-            (*wait 5 seconds*)
-            (* try again *)
-            test (i + 1)
-        | e ->
-            (* ERROR: anything else means that the server is NOT responding adequately *)
-            debug
-              "Request %i/%i to external authentication server failed, waiting \
-               5 secs to try again: %s"
-              i max
+        | () ->
+            true
+        | exception Not_found ->
+            true
+        | exception e ->
+            debug "Failed to remove user %s from cache: %s" full_username
               (ExnHelper.string_of_exn e) ;
-            Thread.delay 5.0 ;
-            (*wait 5 seconds*)
-            (* try again *)
-            test (i + 1)
+            false
+      )
     in
-    debug "Testing if external authentication server is accepting requests..." ;
-    let full_username = get_full_subject_name username in
-    ( try
-        ignore
-          (pbis_common "/opt/pbis/bin/ad-cache"
-             ["--delete-user"; "--name"; full_username])
+    let try_fetch_sid () =
+      try
+        let sid = get_subject_identifier krbtgt in
+        debug
+          "Request to external authentication server successful: user %s was \
+           found"
+          krbtgt ;
+        let (_ : (string * string) list) = query_subject_information sid in
+        debug
+          "Request to external authentication server successful: sid %s was \
+           found"
+          sid ;
+        true
       with
-    | Not_found ->
-        ()
-    | e ->
-        debug "Failed to remove user %s from cache: %s" full_username
-          (ExnHelper.string_of_exn e) ;
-        raise e
-    ) ;
-    test 0
+      | Not_found ->
+          (* that means that pbis is responding to at least cached subject queries.
+             in this case, KRBTGT wasn't found in the AD domain. this usually indicates that the
+             AD domain is offline/inaccessible to pbis, which will cause problems, specially
+             to the ssh python hook-script, so we need to try again until KRBTGT is found, indicating
+             that the domain is online and accessible to pbis queries *)
+          debug
+            "Request to external authentication server returned KRBTGT \
+             Not_found" ;
+          false
+      | e ->
+          debug
+            "Request to external authentication server failed for reason: %s"
+            (ExnHelper.string_of_exn e) ;
+          false
+    in
+    (* in order to properly test that PBIS and the AD domain
+       are running correctly, we first remove KRBTGT from the
+       cache. if we didn't, then in the case that PBIS is running
+       and the AD domain is offline, [ can_fetch_sid ] will
+       erroneously return success *)
+    (* to avoid clearing the cache twice, we carry around a checkpoint *)
+    let rec pbis_functioning = function
```

Comment:
Is this recursive function not just a sequence of two tests?

Reply:
Yes there are only two operations, the idea is to avoid calling `try_clear_cache` twice (though the code would still work if we did)